### DISCUSSION TITLE
'Unchecked runtime.lastError: The message port closed before a response was received' chrome issue?

### DISCUSSION THREAD

**Discussion by newtonfav** [open-sauced-craftwork/browser-extension#26] [Q&A] [answered]
I am trying to send a message from my background.js file to the contentScript file, but it seems not to be working, I am getting 'Unchecked runtime.lastError: The message port closed before a response was received' 

Manifest.json file
`{
  "manifest_version": 3,
  "name": "Opensauced",
  "description": "Detects if github user is signed up to opensaucesd",
  "version": "1.0.0",
  "permissions": ["storage", "tabs"],
  "background": {
    "scripts": ["background.js"],
    "persistent": false
  },
  "content_scripts": [
    {
      "matches": ["https://github.com/*"],
      "js": ["contentScript.js"]
    }
  ],
  "host_permissions": [
    "https://github.com/*",
    "https://api.opensauced.pizza/*"
  ],
  "action": { "default_popup": "index.html" },
  "icons": {
    "16": "icons/icon16.png",
    "32": "icons/icon32.png",
    "48": "icons/icon48.png",
    "128": "icons/icon128.png"
  }
}
`

background.js file
`chrome.runtime.sendMessage({ text: 'hey' }, function (response) {
  console.log('Response: ', response);
});
`
contentScript.js file
`chrome.runtime.onMessage.addListener(function (msg, sender, sendResponse) {
  console.log(msg);
  sendResponse('Gotcha!');
  return true;
});
`

**Comment by anush008** [comment, marked as answer]
Since you are sending a message from a background worker to a content script, you'll have to do it like so:
`
chrome.tabs.query({active: true, currentWindow: true}, function(tabs) {
  chrome.tabs.sendMessage(tabs[0].id, {message: "From BG!"});
});
`
In your content script:
`
chrome.runtime.onMessage.addListener(function(request, sender, sendResponse) {
  console.log(request.message); 
});
`

**Reply by anush008** [reply]
In my project, I sent the the request from the content-script to the background worker and then returned a response from the background worker. 
That way I didn't need the tabs permission.

In my content-script:
`
chrome.runtime.sendMessage({message: "From content-worker"}, function(response) {
  console.log(response.fromBG);
});
`
In the BG worker:
`
chrome.runtime.onMessage.addListener(function(request, sender, sendResponse) {
  sendResponse({fromBG: "From BG!"});
});
`

**Reply by newtonfav** [reply]
If you did yours like this how did you get access to chrome.tabs to extract users current url as chrome.tabs api don't work in the content-script according to chrome extension docs.

**Reply by anush008** [reply]
`document.location` in a content-script returns the location object of the webpage the script is configured for.

**Reply by newtonfav** [reply]
That's a brilliant way to bypass using chrome.tabs, thanks.

**Comment by newtonfav** [comment]
it seems the background script usually unable to communicate with content script even when you return true as response as documented on MDN, a preferable solution is to send message from content script to background instead of vice-versa, but now using document.location to extract url as suggested by @Anush008
